modules/decima-ext-model-exporter/src/main/java/com/shade/decima/ui/data/viewer/model/dmf/DMFExporter.java

```diff
@@ -91,9 +122,109 @@ public void export(
         @NotNull Writer writer
     ) throws Exception {
         final var scene = export(monitor, core, object, resourceName);
+        nodesRemoved = 0;
+        nodesVisited = 0;
+        for (DMFNode node : scene.models) {
+            countNodes(node);
+        }
+        for (DMFNode node : scene.instances) {
+            countNodes(node);
+        }
+        for (DMFNode node : scene.instances) {
+            removeEmpty(node);
+        }
+        optimizeModelList(scene.instances);
+        optimizeInstances(scene);
+        for (DMFNode node : scene.instances) {
+            optimize(node);
+        }
+        optimizeInstances(scene);
+        optimizeModelList(scene.instances);
+        optimizeModelList(scene.models);
+        for (DMFNode node : scene.models) {
+            removeEmpty(node);
+        }
+        for (DMFNode node : scene.models) {
+            optimize(node);
+        }
+        optimizeModelList(scene.models);
```

Comment:
Please use visitor pattern or at least extract it somewhere else. Optimization is not a part of the exporter.